core/src/sumeragi/fault.rs

```diff
@@ -38,866 +37,1337 @@ impl FaultInjection for NoFault {
+            view_change_index: current_view_change_index,
+            signatures: Vec::new(),
+        };
+        proof
+            .sign(sumeragi.key_pair.clone())
+            .expect("Suspect proof must be able to perform signing");
+        proof
+    };
+
+    let _ = view_change_proof_chain.insert_proof(
+        &state_machine
+            .current_topology
+            .sorted_peers()
+            .iter()
+            .cloned()
+            .collect(),
+        state_machine.current_topology.max_faults(),
+        &state_machine.latest_block_hash,
+        &suspect_proof,
+    );
+
+    sumeragi.broadcast_packet(
+        MessagePacket::new(
+            view_change_proof_chain.clone(),
+            Message::ViewChangeSuggested,
+        ),
+        &state_machine.current_topology,
+    );
+}
 
-    /// Get peers as a hash set of their ids.
-    pub fn peers(&self) -> HashSet<PeerId> {
-        self.topology.sorted_peers().iter().cloned().collect()
-    }
+#[instrument(skip(sumeragi, state_machine_guard))]
+#[allow(clippy::expect_used)]
+/// Execute the main loop of [`SumeragiWithFault`]
+pub fn run_sumeragi_main_loop<F>(
+    sumeragi: &SumeragiWithFault<F>,
+    mut state_machine_guard: SumeragiStateMachineData,
+    mut shutdown_receiver: tokio::sync::oneshot::Receiver<()>,
+) where
+    F: FaultInjection,
+{
+    let mut incoming_message_receiver = sumeragi
+        .incoming_message_receiver
+        .lock()
+        .expect("lock on reciever");
 
-    /// Assumes this peer is a leader and starts the round with the given `genesis_topology`.
-    ///
-    /// # Errors
-    /// Can fail if:
-    /// * transactions are empty
-    /// * peer is not leader
-    /// * there are already some blocks in blockchain
-    #[iroha_futures::telemetry_future]
-    #[log(skip(self, transactions, genesis_topology, ctx))]
-    pub async fn start_genesis_round(
-        &mut self,
-        transactions: Vec<VersionedAcceptedTransaction>,
-        genesis_topology: Topology,
-        ctx: &mut Context<Self>,
-    ) -> Result<()> {
-        if transactions.is_empty() {
-            Err(eyre!(
-                "Genesis transactions set contains no valid transactions"
-            ))
-        } else if genesis_topology.leader() != &self.peer_id {
-            Err(eyre!(
-                "Incorrect network topology this peer should be {:?} but is {:?}",
-                Role::Leader,
-                genesis_topology.role(&self.peer_id)
-            ))
-        } else if self.block_height > 0 {
-            Err(eyre!(
-                "Block height should be 0 for genesis round. But it is: {}",
-                self.block_height
-            ))
+    // Initialization code
+    {
+        if state_machine_guard.latest_block_height != 0
+            && state_machine_guard.latest_block_hash != Hash::zeroed().typed()
+        {
+            // Normal startup
+            // Aka, we don't have to do anything.
```

Comment:
Empty branch seems a bit strange, i suppose we could just write comment `Otherwise do nothing (normal startup)`.

Reply:
Another anti-pattern is scope-around-scope. A function call cannot leak internal state. If you don't have a top-level let-binding the scope does nothing.

Reply:
Also, this is the one case, where a short-circuiting `if` statement is the preferred solution.